File: ios/stereoCamera/stereoCamera/Controllers/MasterShake.swift
//
//  MasterShake.swift
//  stereoCamera
//
//  Created by hallmarklabs on 6/24/18.
//  Copyright © 2018 cody. All rights reserved.
//

import Foundation

class MasterShake
{
    private var steps = [MasterShakeStep]()
    private let lock = NSCondition()
    private var handshaking = false;
    private var finalListener:MasterShakeListener?
    private var stepListener:MasterShakeListener?
    private var currentStep = -1
    public var parent:CameraMasterCtrl? = nil
    
    init()
    {
        steps.append(PingStep())
        steps.append(HandshakeStep())
        steps.append(VersionCheckStep())
        steps.append(IDStep())
        steps.append(SetCameraStep(parent: self))
        steps.append(SetQualityStep())
        steps.append(SetZoomStep(parent: self))
        steps.append(SetOverlayStep())
        steps.append(ReadyStep(parent: self))
    }
    
    class StepListener : CommandResponseListener
    {
        var received:(Command, Command?) -> Void
        var failed:() -> Void
        
        init(received: @escaping (Command, Command?) -> Void, failed: @escaping () -> Void)
        {
            self.received = received
            self.failed = failed
        }

        override func onReceived(_ command: Command, origCommand: Command?)
        {
            received(command, origCommand)
        }
        
        override func onReceiveFailed(_ command: Command)
        {
            failed()
        }
    }
    
    class DefaultStepListener : CommandResponseListener
    {
        var listener:MasterShakeListener
        
        init(_ listener:@escaping MasterShakeListener)
        {
            self.listener = listener
        }
        
        override func onReceived(_ command: Command, origCommand: Command?)
        {
            listener(true)
        }
        
        override func onReceiveFailed(_ command: Command)
        {
            listener(false)
        }
    }
    
    class PingStep : MasterShakeStep
    {
        var attempt:Int = 0
    
        var name:String
        {
            get { return "ping" }
        }
        
        func execute(listener: @escaping MasterShakeListener)
        {
            let cmd = Ping()
            CommManager.instance.comm.sendCommand(command: cmd, listener: StepListener(
                received: { (command:Command, origCmd:Command?) in
                    listener(true)
                },
                failed: { [unowned self] in
                    self.attempt += 1
            
                    if (self.attempt == 3)
                        { listener (false) }
                    else
                    {
                        Thread.sleep(forTimeInterval: 1000)
                        self.execute(listener: listener)
                    }
                })
            )
        }
    }
    
    class VersionCheckStep : MasterShakeStep
    {
        var name: String { get { return "version check" }}
        
        func execute(listener: @escaping MasterShakeListener)
        {
            let cmd = Version()
            CommManager.instance.comm.sendCommand(command: cmd, listener: StepListener(
                received: {(_ respCmd: Command, _ origCmd: Command?) -> Void
                in
                    let orig = origCmd as! Version
                    let resp = respCmd as! Version
                    
                    Cookie.instance.setCurrentClientVersion(value: resp.version)
                    Cookie.instance.setCurrentClientPlatform(value: resp.platform)
                    
                    if (resp.platform == .IOS && orig.version == resp.version)
                        { listener(true) }
                    else if (resp.platform == .ANDROID && resp.version == 1200)
                        { listener(true) }
                    else
                        { listener(false) }
                },
                failed: {() -> Void
                in
                    listener(false)
                }
            ))
        }
    }

    class HandshakeStep : MasterShakeStep
    {
        var name:String
        {
            get { return "handshake" }
        }
        
        func execute(listener: @escaping MasterShakeListener)
        {
            let cmd = Handshake()
            CommManager.instance.comm.sendCommand(command: cmd, listener: DefaultStepListener(listener))
        }
    }
    
    class IDStep : MasterShakeStep
    {
        var name:String
        {
            get { return "id" }
        }
        
        func execute(listener: @escaping MasterShakeListener)
        {
            let cmd = ID(phoneId: Cookie.instance.id)
            CommManager.instance.comm.sendCommand(command: cmd, listener: StepListener(
                received:
                { (_ respCmd: Command, _ origCmd: Command?) -> Void in
                    let cmd:ID = respCmd as! ID
                    Cookie.instance.setCurrentClientID(value: cmd.phoneId)
                    listener(true)
                },
                failed:
                {() -> Void in
                    listener(false)
                }
            ))
        }
    }

    class ReadyStep : MasterShakeStep
    {
        let parent:MasterShake
        init(parent: MasterShake)
        {
            self.parent = parent
        }
    
        var name:String
        {
            get { return "ready"}
        }
        
        func execute(listener: @escaping MasterShakeListener)
        {
            let statusList = [Status.LISTENING, Status.READY]
            parent.parent?.slaveState.waitOnStatusAsync(statusList: statusList, timeout: 3.0, listener: {(success: Bool) -> Void
            in
                listener(success)
            })
        }
    }
    
    class CreatedStep : MasterShakeStep
    {
        let parent:MasterShake
        init(parent: MasterShake)
        {
            self.parent = parent
        }
    
        var name:String { get { return "created"} }
        
        func execute(listener: @escaping MasterShakeListener)
        {
            let statusList = [Status.RESUMED, Status.CREATED]
            parent.parent?.slaveState.waitOnStatusAsync(statusList: statusList, timeout: 3.0, listener: {(success: Bool) -> Void
            in
                listener(success)
            })
        }
    }
    
    class SetCameraStep : MasterShakeStep
    {
        unowned let parent:MasterShake
        init(parent: MasterShake)
        {
            self.parent = parent
        }
        var name:String { get { return "set camera" }}
        
        func execute(listener: @escaping MasterShakeListener)
        {
            let position = parent.parent?.cameraPreview.currentCamera?.position
            let posBool = (position != nil && position == .front) ? true : false
            
            let cmd = SetFacing(posBool)
            CommManager.instance.comm.sendCommand(command: cmd, listener: DefaultStepListener(listener))
        }
    }
    
    class SetZoomStep : MasterShakeStep
    {
        unowned let parent:MasterShake
        init(parent: MasterShake)
        {
            self.parent = parent
        }
        var name:String { get { return "set zoom" }}
        
        func execute(listener: @escaping MasterShakeListener)
        {
            let zoom = Cookie.instance.getZoomForClient(isMaster: false, client: CommManager.instance.comm.address, camera: (parent.parent?.cameraPreview.currentCamera?.position)!)
            let cmd = SetZoom(localZoom: 1.0, remoteZoom: zoom)
            CommManager.instance.comm.sendCommand(command: cmd, listener: DefaultStepListener(listener))
        }
    }
    
    class SetOverlayStep : MasterShakeStep
    {
        var name:String { get { return "set overlay" }}
        
        func execute(listener: @escaping MasterShakeListener)
        {
            let overlay = Cookie.instance.overlay
            let cmd = SetOverlay(overlay)
            CommManager.instance.comm.sendCommand(command: cmd, listener: DefaultStepListener(listener))
        }
    }
    
    class SetQualityStep : MasterShakeStep
    {
        var name: String { get { return "set image quality" }}
        
        func execute(listener: @escaping MasterShakeListener)
        {
            let quality = Cookie.instance.imageQuality
            let cmd = SetCaptureQuality(quality)
            CommManager.instance.comm.sendCommand(command: cmd, listener: DefaultStepListener(listener))
        }
    }
    
    func start(listener: @escaping MasterShakeListener)
    {
        var doReturn = false
        lock.lock()
            if (handshaking)
                { doReturn = true }
            else
                { handshaking = true }
        lock.unlock()
        
        if (doReturn)
            { return }
        
        finalListener = listener
        stepListener = {(success:Bool) ->Void
        in
            if (success)
            {
                print ("step success")
                self.nextStep()
            }
            else
            {
                print ("step fail")
                self.runFail()
            }
        }
        
        currentStep = -1
        nextStep();
    }
    
    private func nextStep()
    {
        currentStep += 1
        
        if (currentStep == steps.count)
        {
            lock.lock()
                handshaking = false
            lock.unlock()
            
            print ("handshake success")
            finalListener?(true)
            return
        }
        
        let step = steps[currentStep]
        print("executing handshake step " + String(currentStep) + ": " + step.name)
        
        if (stepListener != nil)
            {step.execute(listener: stepListener!)}
        else
            {finalListener?(false)}
    }
    
    private func runFail()
    {
        lock.lock()
            handshaking = false
        lock.unlock()
        
        finalListener?(false)
    }
}

typealias MasterShakeListener = (_ success:Bool) -> Void

protocol MasterShakeStep
{
    var name:String {get}
    func execute(listener: @escaping MasterShakeListener)
}

File: android/app/src/main/java/com/munger/stereocamera/fragment/MasterShake.java
package com.munger.stereocamera.fragment;

import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.util.Log;

import com.munger.stereocamera.MyApplication;
import com.munger.stereocamera.ip.command.CommCtrl;
import com.munger.stereocamera.ip.command.commands.Handshake;
import com.munger.stereocamera.ip.command.commands.ID;
import com.munger.stereocamera.ip.command.commands.Ping;
import com.munger.stereocamera.ip.command.commands.SetCaptureQuality;
import com.munger.stereocamera.ip.command.commands.SetFacing;
import com.munger.stereocamera.ip.command.commands.SetOverlay;
import com.munger.stereocamera.ip.command.commands.SetZoom;
import com.munger.stereocamera.ip.command.commands.Version;
import com.munger.stereocamera.ip.utility.RemoteState;
import com.munger.stereocamera.utility.data.ClientViewModel;
import com.munger.stereocamera.widget.PreviewOverlayWidget;
import com.munger.stereocamera.widget.PreviewWidget;

import java.util.ArrayList;

public class MasterShake
{
	private boolean handshaking = false;
	private int currentStep = 0;
	private final Object lock = new Object();
	private Listener finalListener;
	private StepListener stepListener;

	private final long PING_TIMEOUT = 3000;
	private final long HANDSHAKE_TIMEOUT = 1000;
	private final long LISTEN_STATUS_TIMEOUT = 10000;
	private final long READY_STATUS_TIMEOUT = 6000;

	private final MasterFragment target;
	private final ClientViewModel.CameraDataPair cameraPair;
	private CommCtrl comm;

	public MasterShake(MasterFragment target, ClientViewModel.CameraDataPair cameraPair)
	{
		Log.d("stereoCamera", "new handshake created");

		this.target = target;
		this.cameraPair = cameraPair;

		init();

		steps.add(pingStep);
		steps.add(handshakeStep);
		steps.add(versionCheckStep);
		steps.add(idStep);
		steps.add(setCameraStep);
		steps.add(setQualityStep);
		steps.add(setZoomStep);
		steps.add(overlayStep);
		steps.add(readyStep);
	}

	public void start(Listener listener)
	{
		synchronized (lock)
		{
			if (handshaking)
				return;

			handshaking = true;
		}

		finalListener = listener;
		stepListener = new StepListener()
		{
			@Override
			public void success()
			{
				Log.d("stereoCamera", "step success");
				nextStep();
			}

			@Override
			public void fail()
			{
				Log.d("stereoCamera", "step fail");
				runFail();
			}
		};

		currentStep = -1;
		nextStep();
	}

	private void nextStep()
	{
		currentStep++;

		if (currentStep == steps.size())
		{
			synchronized (lock)
			{
				handshaking = false;
			}

			Log.d("stereoCamera", "handshake success");
			finalListener.success();
			return;
		}

		Step step = steps.get(currentStep);
		Log.d("stereoCamera", "executing handshake step " + currentStep + " : " + step.name);
		step.execute(stepListener);
	}

	private void runFail()
	{
		synchronized (lock)
		{
			handshaking = false;
		}

		finalListener.fail();
	}

	private final ArrayList<Step> steps = new ArrayList<>();

	public interface Listener
	{
		void success();
		void fail();
	}

	private interface StepListener
	{
		void success();
		void fail();
	}

	private abstract static class Step
	{
		public String name;

		public Step(String name)
		{
			this.name = name;
		}

		public abstract void execute(StepListener listener);
	}

	private String getTag()
	{
		return "Master shake";
	}

	private Step pingStep;
	private Step versionCheckStep;
	private Step setCameraStep;
	private Step overlayStep;
	private Step handshakeStep;
	private Step idStep;
	private Step setQualityStep;
	private Step setZoomStep;
	private Step readyStep;

	private void init()
	{
		comm = MyApplication.getInstance().getCtrl();

		pingStep = new Step("ping") { public void execute(final StepListener listener)
		{
			Log.d("stereoCamera", "pinging slave phone");
			comm.sendCommand(new Ping(), new CommCtrl.DefaultResponseListener((success, command, originalCmd) -> {
				if (success)
				{
					Log.d("stereoCamera", "ping success");
					listener.success();
				}
				else
					listener.fail();
			}), PING_TIMEOUT);
		}};

		handshakeStep = new Step("handshake") { public void execute(final StepListener listener)
		{
			Log.d("stereoCamera", "initiating handshake");
			comm.sendCommand(new Handshake(), new CommCtrl.DefaultResponseListener((success, command, originalCmd) -> {
			if (success)
			{
				Log.d("stereoCamera", "handshake started");
				listener.success();
			}
			else
				listener.fail();
		}), HANDSHAKE_TIMEOUT);
		}};

		idStep = new Step("id") { public void execute(final StepListener listener)
		{
			Log.d("stereoCamera", "fetching id");
			comm.sendCommand(new ID(), new CommCtrl.DefaultResponseListener((success, command, originalCmd) -> {
			if (success)
			{
				ID cmd = (ID) command;
				Log.d("stereoCamera", "remote camera ID " + cmd.phoneId);
				listener.success();
			}
			else
				listener.fail();
		}), 3000);
		}};

		setQualityStep = new Step("quality") { public void execute(final StepListener listener)
		{
			Log.d("stereoCamera", "setting image quality");

			SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(MyApplication.getInstance());
			String captureTypeStr = sharedPref.getString("pref_capture", "preview");
			PreviewWidget.ShutterType captureType;

			switch(captureTypeStr)
			{
				case "hi cam":
					captureType = PreviewWidget.ShutterType.HI_RES;
					break;
				case "lo cam":
					captureType = PreviewWidget.ShutterType.LO_RES;
					break;
				default:
					captureType = PreviewWidget.ShutterType.PREVIEW;
					break;
			}

			comm.sendCommand(new SetCaptureQuality(captureType), new CommCtrl.DefaultResponseListener((success, command, originalCmd) -> {
			if (success)
				listener.success();
			else
				listener.fail();
			}), HANDSHAKE_TIMEOUT);
		}};

		setZoomStep = new Step("zoom") { public void execute(final StepListener listener) {
			Log.d("stereoCamera", "setting zoom value");
			SetZoom cmd = new SetZoom(cameraPair.local.zoom, cameraPair.remote.zoom);
			comm.sendCommand(cmd, new CommCtrl.DefaultResponseListener((success, command, originalCmd) -> {
				if (success)
					listener.success();
				else
					listener.fail();
			}), HANDSHAKE_TIMEOUT);
		}};

		readyStep = new Step("listen ready") { public void execute(final StepListener listener)
			{
				PreviewFragment.Status[] statuss = new PreviewFragment.Status[]{PreviewFragment.Status.LISTENING, PreviewFragment.Status.READY};
				target.remoteState.waitOnStatusAsync(statuss, LISTEN_STATUS_TIMEOUT, new RemoteState.ReadyListener()
				{
					@Override
					public void done()
					{
						listener.success();
					}

					@Override
					public void fail()
					{
						listener.fail();
					}
				});
			}
		};

		versionCheckStep = new Step("check version") { public void execute(final StepListener listener)
		{
			comm.sendCommand(new Version(), new CommCtrl.DefaultResponseListener((success, command, originalCmd) -> {
				if (!success)
				{
					listener.fail();
					return;
				}

				Version resp = (Version) command;
				int remoteVersion = resp.version;

				if (resp.platform == Version.Platform.ANDROID && remoteVersion > ((Version)originalCmd).version)
					listener.fail();
				else if (resp.platform == Version.Platform.IOS && remoteVersion > 1000)
					listener.fail();
				else
					listener.success();
				}), 1000);
		}};

		class SetupCameraStep extends Step
		{
			private boolean localFacing = false;
			private boolean remoteFacing = false;
			private final Object facingLock = new Object();
			private boolean listenerCalled = false;
			private long start;
			private StepListener listener;

			public SetupCameraStep(String name)
			{
				super(name);
			}

			public void execute(final StepListener listener)
			{
				start = System.currentTimeMillis();
				this.listener = listener;
				localFacing = false;
				remoteFacing = false;

				boolean isFacing = cameraPair.local.isFacing;
				target.startPreview();
				target.setCamera(isFacing, new MasterFragment.SetCameraListener()
				{
					@Override
					public void done()
					{
						Log.d("stereoCamera", "set local camera finished");
						synchronized (facingLock)
						{
							localFacing = true;
						}

						execute2();
					}

					@Override
					public void fail()
					{
						Log.d("stereoCamera", "set local camera failed");
					}
				});

				comm.sendCommand(new SetFacing(isFacing), new CommCtrl.DefaultResponseListener((success, command, originalCmd) -> {
					Log.d("stereoCamera", "set remote camera finished");
					if (success)
					{
						synchronized (facingLock)
						{
							remoteFacing = true;
						}

						execute2();
					}
				}), 3000);

				Thread timeoutThread = new Thread(() -> {
					synchronized (facingLock)
					{
						if (!listenerCalled)
						{
							try{facingLock.wait(3000);} catch(InterruptedException ignored){}
						}

						if (!listenerCalled)
						{
							execute2();
						}
					}
				});
				timeoutThread.start();
			}

			private void execute2()
			{
				boolean doSuccess = false;
				boolean doFail = false;
				synchronized (facingLock)
				{
					if (listenerCalled)
						return;

					if (localFacing && remoteFacing)
					{
						doSuccess = true;
						listenerCalled = true;
					}
					else if (System.currentTimeMillis() - start > 3000)
					{
						doFail = true;
						listenerCalled = true;
					}

					if (listenerCalled)
						facingLock.notify();
				}

				if (doSuccess)
					listener.success();
				else if (doFail)
					listener.fail();
			}
		}

		setCameraStep = new SetupCameraStep("setup camera");

		overlayStep = new Step("overlay") { public void execute(final StepListener listener)
		{
			final PreviewOverlayWidget.Type type = target.overlayWidget.getType();
			target.masterComm.sendCommand(new SetOverlay(type), (success, command, originalCmd) -> {
				if (success)
				{
					if (type == PreviewOverlayWidget.Type.Ghost)
						target.slavePreview.start();

					listener.success();
				}
				else
					listener.fail();
			});
		}};
	}
}
